### PR TITLE
(plugin) apps::whois::cli - add new plugin to monitor domain expiry date

### PR DESCRIPTION
**Pull request by PedroMSantosD**
This pull request adds a module to monitor domain expiration date using Linux tool: /usr/sbin/whois and 
a pattern for the expiry date, that identifies regex groups for year, month, day and optional hour minute second on the whois response.

### CODE REVIEW

**Comment by PedroMSantosD** [comment]
linked to https://github.com/centreon/centreon-plugins/issues/1135

**Comment by Sims24** [comment]
Hi @PedroMSantosD 

Many thanks for this first contribution, we always feel good when someone takes the time to get their hands dirty to fix an issue. The regex pattern is a smart solution. 

I see that the code is working just fine but we need to ensure that the codebase remains as standard as possible. To merge your code I should ask you to modify your code to fit our standards. 

We will provide a detailed review with examples. If you have any question, just reach us on slack through centreon #centreon-plugins dedicated channel!

**Comment by PedroMSantosD** [comment]
Hi @Sims24 , Thanks for the feedback. Kindly let me know what needs to be done so the code is compliant with your standards.
Best,

**Comment by PedroMSantosD** [comment]
Hi @Sims24 , @garnier-quentin  any updates on the review required ?

Best,

File: apps/whois/mode/whois.pm [hunk 1]
```diff
@@ -0,0 +1,308 @@
...
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+# Authors : Pedro Manuel Santos Delgado - lunik
+#
+
+package apps::whois::mode::whois;
```
Comment (Sims24) [inline apps/whois/mode/whois.pm:23]: please use apps::whois::cli::mode::whois

File: apps/whois/mode/whois.pm [hunk 2]
```diff
@@ -0,0 +1,308 @@
...
+    $options{options}->add_options(arguments => {
+        "domain:s"            => { name => 'domain'},
+        "whois_binary:s"      => { name => 'command', default => '/usr/bin/whois'},
+        "timeout:s"           => { name => 'timeout', default => 3},
+        "warning-days:s"      => { name => 'warning', default => "30:"},
+        "critical-days:s"     => { name => 'critical', default => "10:"},
+        "custom-date-regex:s" => { name => 'custom_date_regex', default => 'Expiry Date:\s+(\d{2})-(\d{2})-(\d{4})'}, # Regex for whois outcom
+        "custom-date-map:s"   => { name => 'custom_date_map', default => "DMY"}, # day month year
+    });
+    # This method will execute the command, but will not let me parse the outcome
```
Comment (Sims24) [inline apps/whois/mode/whois.pm:45]: Could you explain this comment?

File: apps/whois/mode/whois.pm [hunk 3]
```diff
@@ -0,0 +1,308 @@
...
+    # Check domain has only two levels
+    my $domain_levels = split('\.', $self->{option_results}->{domain});
+    if ($domain_levels != 2){
+        $self->{output}->add_option_msg(
+            short_msg => "Wrong domain pattern '" . $self->{option_results}->{domain} . "'. Only DOMAIN.TLD supported");
+        $self->{output}->option_exit();
+    }
+    # Check custom date regex has at least 3 groups 
+    my $MinGroupCount = 3;
+    my $opening_groups = ($self->{option_results}->{custom_date_regex} =~ tr/\(//);
```
Comment (Sims24) [inline apps/whois/mode/whois.pm:61]: Try to avoid mixing CamelCase and snake_case in scalar names

File: apps/whois/mode/whois.pm [hunk 4]
```diff
@@ -0,0 +1,308 @@
...
+
+
+    if (($self->{perfdata}->threshold_validate(label => 'warning', value => $self->{option_results}->{warning})) == 0) {
+        $self->{output}->add_option_msg(short_msg => "Wrong warning threshold '" . $self->{option_results}->{warning} . "'.");
+        $self->{output}->option_exit();
+    }
+    if (($self->{perfdata}->threshold_validate(label => 'critical', value => $self->{option_results}->{critical})) == 0) {
+        $self->{output}->add_option_msg(short_msg => "Wrong critical threshold '" . $self->{option_results}->{critical} . "'.");
+        $self->{output}->option_exit();
+    }
```
Comment (Sims24) [inline apps/whois/mode/whois.pm:89]: YOu will not need this after refactoring on top of counter template class

File: apps/whois/mode/whois.pm [hunk 5]
```diff
@@ -0,0 +1,308 @@
...
+# limitations under the License.
+#
+# Authors : Pedro Manuel Santos Delgado - lunik
+#
+
+package apps::whois::mode::whois;
+
+use strict;
+use warnings;
+use base qw(centreon::plugins::mode);
```
Comment (Sims24) [inline apps/whois/mode/whois.pm:27]: Prefer using `use base qw(centreon::plugins::templates::counter);` 

See here https://github.com/centreon/centreon-plugins/blob/8ea8c2b7cbf143b40fa5b6110d7ab78fa3679fed/hardware/server/ibm/hmc/ssh/mode/ledstatus.pm#L23

File: apps/whois/plugin.pm [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+# Authors : Pedro Manuel Santos Delgado - lunik
+#
+package apps::whois::plugin;
+
+use strict;
+use warnings;
+use base qw(centreon::plugins::script_simple);
```
Comment (Sims24) [inline apps/whois/plugin.pm:26]: Use script_custom here. 

See: https://github.com/centreon/centreon-plugins/blob/5216525e8b440a1b40c4ab5ab5e811e6c3dd8974/apps/nmap/cli/plugin.pm#L25

File: apps/whois/plugin.pm [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+    my ($class, %options) = @_;
+
+    my $self = $class->SUPER::new(package => __PACKAGE__, %options);
+    bless $self, $class;
+
+    $self->{version} = '1.0';
+    $self->{modes} = {
+        'whois'               => 'apps::whois::mode::whois'
+    };
+
```
Comment (Sims24) [inline apps/whois/plugin.pm:38]: Use the custom cli mode which is the last standard. 

See: https://github.com/centreon/centreon-plugins/blob/5216525e8b440a1b40c4ab5ab5e811e6c3dd8974/apps/nmap/cli/plugin.pm#L37

File: apps/whois/plugin.pm [hunk 8]
```diff
@@ -0,0 +1,51 @@
...
+sub new {
+    my ($class, %options) = @_;
+
+    my $self = $class->SUPER::new(package => __PACKAGE__, %options);
+    bless $self, $class;
+
+    $self->{version} = '1.0';
+    $self->{modes} = {
+        'whois'               => 'apps::whois::mode::whois'
```
Comment (Sims24) [inline apps/whois/plugin.pm:36]: Don't forget to add ::cli:: here as well ;)

File: apps/whois/requirements [hunk 9]
```diff
@@ -0,0 +1,2 @@
+perl-DateTime.x86_64 : Date and time object
+whois-5.1.1-2.el7.x86_64  : /usr/bin/whois binary
```
Comment (Sims24) [inline apps/whois/requirements:2]: Please add it to the PLUGIN DESCRIPTION if you feel like it but don't create a dedicated file for it.

File: apps/whois/mode/whois.pm [hunk 10]
```diff
@@ -0,0 +1,308 @@
...
+        "timeout:s"           => { name => 'timeout', default => 3},
+        "warning-days:s"      => { name => 'warning', default => "30:"},
+        "critical-days:s"     => { name => 'critical', default => "10:"},
+        "custom-date-regex:s" => { name => 'custom_date_regex', default => 'Expiry Date:\s+(\d{2})-(\d{2})-(\d{4})'}, # Regex for whois outcom
+        "custom-date-map:s"   => { name => 'custom_date_map', default => "DMY"}, # day month year
+    });
+    # This method will execute the command, but will not let me parse the outcome
+    return $self;
+}
+
```
Comment (Sims24) [inline apps/whois/mode/whois.pm:48]: Add your counters definition here. 

For example you'll have a global/days counter, type 0. 

See a full example of something similar here: https://github.com/centreon/centreon-plugins/blob/master/apps/protocols/x509/mode/certificate.pm

File: apps/whois/mode/whois.pm [hunk 11]
```diff
@@ -0,0 +1,308 @@
...
+        $self->{output}->add_option_msg(short_msg => "Wrong critical threshold '" . $self->{option_results}->{critical} . "'.");
+        $self->{output}->option_exit();
+    }
+}
+
+sub run {
+    my ($self, %options) = @_;
+    my $catch_error_message = '';
+    # Extracts expiry date
+    my ($lerror, $stdout, $exit_code) = centreon::plugins::misc::backtick(
```
Comment (Sims24) [inline apps/whois/mode/whois.pm:96]: Please try to use the execute_command from the custom_mode template like we do here: https://github.com/centreon/centreon-plugins/blob/2fd7bbb451f3dbf44b193d3566649bd0c2ed8c3e/os/linux/local/mode/uptime.pm#L60

File: apps/whois/mode/whois.pm [hunk 12]
```diff
@@ -0,0 +1,308 @@
...
+        # Exit error unable to get timestamp.  
+        } else {
+            $catch_error_message = 'Whois response did not match known Expire Regex'; 
+            $self->{output}->output_add(
+                severity => 'Unknown',
+                short_msg => $catch_error_message,
+                long_msg => $long_error_message
+                );
+        }
+    }else{
```
Comment (Sims24) [inline apps/whois/mode/whois.pm:144]: space between keywords and braces

File: apps/whois/mode/whois.pm [hunk 13]
```diff
@@ -0,0 +1,308 @@
...
+        # Error calling binary 
+        $catch_error_message = "Uncaught error calling binary";
+        $self->{output}->output_add(
```
Comment (Sims24) [inline apps/whois/mode/whois.pm:147]: output will be handled by your counter definition

**Review by Sims24 — CHANGES_REQUESTED** [review]
Here is a first batch of feedbacks, I probably forgot a lot of things but it would be a good start. 

Thanks to help us improve centreon_plugins scope.

File: apps/whois/mode/whois.pm (hunk 1 above)
Comment (PedroMSantosD) [reply apps/whois/mode/whois.pm:23]: done

File: apps/whois/mode/whois.pm (hunk 2 above)
Comment (PedroMSantosD) [reply apps/whois/mode/whois.pm:45]: removed.

File: apps/whois/mode/whois.pm (hunk 3 above)
Comment (PedroMSantosD) [reply apps/whois/mode/whois.pm:61]: CamelCase Removed

File: apps/whois/requirements (hunk 9 above)
Comment (PedroMSantosD) [reply apps/whois/requirements:2]: done.

**Comment by garnier-quentin** [comment]
Sorry but it's missing a lot.

You can check that nice following plugin: https://github.com/HariSekhon/Nagios-Plugins/blob/master/check_whois.pl

**Comment by PedroMSantosD** [comment]
still in progress, It's been quite hectic on my end.

**Comment by PedroMSantosD** [comment]
Hi @Sims24 , I have tried to implement the class 'centreon::plugins::templates::counter' as referred o other plugins, but can't find a way to properly execute the bash command to call the whois binary; can you help point out what is missing?

this error i get on trying to execute de plugin, but I'm afraid I don't know how to troubleshoot it.
```
$ perl centreon_plugins.pl --plugin apps::whois::plugin  --mode whois --domain google.com --warning-days_to_expiration 5
UNKNOWN: Can't call method "execute_command" on an undefined value at /home/pesantos/Repos/centreon-plugins/apps/whois/cli/mode/whois.pm line 116.

```
Thanks in advance

File: apps/whois/mode/whois.pm (hunk 11 above)
Comment (PedroMSantosD) [reply apps/whois/mode/whois.pm]:
```
UNKNOWN: Can't call method "execute_command" on an undefined value at /home/pesantos/Repos/centreon-plugins/apps/whois/cli/mode/whois.pm line 116.
```
help sought...

File: apps/whois/mode/whois.pm (hunk 4 above)
Comment (PedroMSantosD) [reply apps/whois/mode/whois.pm]: removed

File: apps/whois/plugin.pm (hunk 7 above)
Comment (Sims24) [reply apps/whois/plugin.pm:38]: @PedroMSantosD look at this comment this is why the option{custom}->execute_command fails

**Comment by Sims24** [comment]
Hi, thanks for your work to stick with the best practices.

I added a comment I think you don't define the custom mode in the plugin.pm that's why it results in a undefined value.

**Comment by PedroMSantosD** [comment]
Hi @Sims24 , apologies for being off this pull request, I´m having a hard time using the custom plugin, as the invocation to whois binary has an exit code 1, a I have not been able to catch it on an eval section, as the classes just crash, ending the execution.

Have you come with such a situation, where you need to catch en error code... which is expeted anyway?
if so, how can you prevent the error from breaking the classes?

Thanks in advance

**Pull request closed without merging**